Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,735 @@
+```
+
+Notice how it gets us what is essentially nested lists, unlike the regular list monad.
+Notice also that it's essentially a rose tree.
+Of course, we can get the standard list monad behavior by concatenating the nested lists, but we might not necessarily want to do that.
+
+In general, we can restore any proper monad behavior from a free monad, since it doesn't define any semantics beyond those necessary for any monad.
+`free` provides a function for that:
+
+```haskell
+foldFree :: Monad m => (forall x. f x -> m x) -> Free f a -> m a
+```
+
+Given a function[^natural-transformation] converting `f x` into some monad `m x` for any `x`, we can convert `Free f a` into `m a`.
+Returning to our list example, we can get the original list behavior by applying `foldFree id`:
+
+```haskell
+> foldFree id listComputation
+[571,671,581,681,572,672,582,682,573,673,583,683]
+```
+
+[^natural-transformation]: This function is a natural transformation.
+    The naturality condition is enforced by parametricity.
+    That being said, the discussion of naturality is not particularly relevant here.
+    Essentially, you can think of any function with type `forall a. f a -> g a` as a natural transformation between `f` and `g`, as long as you ignore `undefined` and other bottoms.
+
+### Free monads for EDSLs
+
+By now, we hopefully got some intuition for how free monads work.
+Let us now define a toy calculator language that reads a few integers from the standard input, adds them together, and prints the result.
+
+The key intuition here is that if the next action depends on the previous one, we need to encode this as a continuation, i.e. we need to have a function accepting some argument and returning the functor parameter.
+
+Now, if we try to define a functor where every operation uses the same type everywhere, we'll run into a bit of an issue.
+Consider this datatype for instance:
+
+```haskell
+data ASTF a
+  = Add a a (a -> a)
+  | Input (a -> a)
+  | Output a a
+```
+
+If we try to derive a `Functor` instance, GHC will complain:
+
+```
+Constructor ‘Add’ must not use the type variable in a function argument
+```
+
+It is, in fact, impossible to define a lawful `Functor` instance if the functor's type variable ends up in the function argument position[^variance].
+Hence, we will introduce two type variables: one for operation arguments, and another for the overall result:
+
+[^variance]: This has to do with `Functor` being a covariant functor, but discussion of variance is out of scope here.
+
+```haskell
+data ASTF t a
+  = Add t t (t -> a)
+  | Input (t -> a)
+  | Output t a
+  deriving Functor
+```
+
+Notice that for `Add` we're saying that it takes two arguments `t`, its result is also `t`, but ultimately, the continuation returns the type `a`.
+In practice, we will define `add :: t -> t -> Free (ASTF t) t`, but to have a lawful `Functor`, and also more flexibility, we're not assuming these types are the same.
+Similar reasoning applies to `Input`.
```

Comment:
I actually can't move it up there, because I didn't at that point introduce the notion that we have to use _two_ type parameters. I can add a short footnote explaining what's going on though.